### PR TITLE
[`flake8-pyi`] Implement `PYI063`

### PR DESCRIPTION
**Pull request by tusharsadhwani**
## Summary
Implements `Y063` from `flake8-pyi`.

## Test Plan
`cargo test` / `cargo insta review`

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    let mut args = function.parameters.args.iter();
+    let Some(first_arg) = args.next() else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+    // TODO: this scope is wrong.
+    let scope = semantic.current_scope();
+    let function_type = function_type::classify(
```
Comment (tusharsadhwani) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs:55]: This call currently doesn't correctly identify methods, which is why the issue is not correctly raised on line 26 of `PYI063.pyi`.

I didn't look super hard into it, but I don't see a very straightforward way to get the correct scope from a `FunctionStmt` node.

File: crates/ruff_linter/src/checkers/ast/analyze/definitions.rs [hunk 2]
```diff
@@ -149,23 +150,22 @@ pub(crate) fn definitions(checker: &mut Checker) {
             if checker.enabled(Rule::DocstringInStub) {
                 flake8_pyi::rules::docstring_in_stubs(checker, docstring);
             }
+            if checker.enabled(Rule::OldStylePositionalOnlyArg) {
+                flake8_pyi::rules::old_style_positional_only_arg(checker, definition);
+            }
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/checkers/ast/analyze/definitions.rs:155]: We should make sure only to run this rule if the target version is set to Python 3.8 or greater (Ruff still supports Python 3.7), or we'll be recommending invalid syntax for Python 3.7 users

File: crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap:1]: Any reason not to do this for `.py` files as well? I think type checkers understand the pre-PEP-570 convention for runtime source code as well as stubs

**Review by AlexWaygood — COMMENTED** [review]
Nice! The code here looks really clean. Charlie's assigned this to himself so I'll leave a full review to him, but two quick points:

File: crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI063.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+def bad(__x: int) -> None: ...  # PYI063
+def also_bad(__x: int, __y: str) -> None: ...  # PYI063
+def still_bad(__x_: int) -> None: ...  # PYI063
+
+def no_args() -> None: ...
+def okay(__x__: int) -> None: ...
+# The first argument isn't positional-only, so logically the second can't be either:
+def also_okay(x: int, __y: str) -> None: ...
+def fine(x: bytes, /) -> None: ...
```
Comment (VascoSch92) [inline crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI063.py:14]: Can I suggest the test case
```python
def still_fine(_x: int) -> None: ...
```
This should be fine right?

File: crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap
Comment (tusharsadhwani) [reply crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap:1]: For regular python files, code like:

```python
def add(__a, __b):
    return __a + __b
```

It's not necessary that the user intended to make them positional only. Whereas in stub files it is intended. Raising it on python files can cause false positives.

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap:1]: Although you certainly _can_ do calls such as `add(__a=1, __b=2)`, in my opinion, I think the vast majority of Python functions with parameters like that have probably been written to take advantage of the type-checker convention to understand those parameters as being positional-only. But let's see what Charlie thinks!

File: crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI063.py (hunk 3 above)
Comment (tusharsadhwani) [reply crates/ruff_linter/resources/test/fixtures/flake8_pyi/PYI063.py:14]: seems good, I'll add it.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            OldStylePositionalOnlyArg,
+            first_arg.range(),
+        ));
+    }
+    if matches!(function_type, function_type::FunctionType::Method) {
+        if let Some(second_arg) = args.next() {
+            if is_old_style_positional_only_arg(second_arg) {
+                checker.diagnostics.push(Diagnostic::new(
+                    OldStylePositionalOnlyArg,
+                    second_arg.range(),
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs:74]: Why is the check limited to the first two arguments?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs]: Ah I see, ok.

**Review by charliermarsh — APPROVED** [review]
Thanks!

File: crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/snapshots/ruff_linter__rules__flake8_pyi__tests__PYI063_PYI063.py.snap:1]: I ended up enabling this based on Alex's suggestion.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/old_style_positional_only_arg.rs]: I fixed this by moving it out of the definition phase and into the standard statement checker.

**Pull request merged**